src/main/java/reposense/report/ReportGenerator.java

```diff
@@ -48,44 +46,57 @@ public static void generateReposReport(List<RepoConfiguration> configs, String o
-            try {
-                GitClone.clone(config);
-                repoReportDirectory = Paths.get(outputPath, config.getDisplayName());
-                FileUtil.createDirectory(repoReportDirectory);
-            } catch (GitCloneException gde) {
-                logger.log(Level.WARNING,
-                        "Exception met while trying to clone the repo, will skip this repo.", gde);
-                repoReportDirectory = Paths.get(outputPath, config.getDisplayName());
-                FileUtil.createDirectory(repoReportDirectory);
-                generateEmptyRepoReport(repoReportDirectory.toString());
-                continue;
-            } catch (IOException ioe) {
-                logger.log(Level.WARNING,
-                        "Error has occurred while creating repo directory, will skip this repo.", ioe);
-                continue;
-            } catch (RuntimeException rte) {
-                logger.log(Level.SEVERE, "Error has occurred during analysis, will skip this repo.", rte);
-                continue;
-            }
+        cloneAndAnalyzeRepos(configs, outputPath);
 
-            // preprocess the config and repo
-            updateRepoConfig(config);
-            updateAuthorList(config);
+        FileUtil.writeJsonFile(new SummaryReportJson(configs, generationDate), getSummaryResultPath(outputPath));
+        logger.info("The report is generated at " + outputPath);
+    }
 
-            CommitContributionSummary commitSummary = CommitsReporter.generateCommitSummary(config);
-            AuthorshipSummary authorshipSummary = AuthorshipReporter.generateAuthorshipSummary(config);
-            generateIndividualRepoReport(commitSummary, authorshipSummary, repoReportDirectory.toString());
+    /**
+     * Clone, analyze and generate the report for repositories in {@code configs}.
+     * Performs analysis and report generation of each repository in parallel with the cloning of the next repository.
+     */
+    private static void cloneAndAnalyzeRepos(List<RepoConfiguration> configs, String outputPath) throws IOException {
+        RepoCloner repoCloner = new RepoCloner();
+        RepoConfiguration clonedRepo = null;
+
+        for (RepoConfiguration config : configs) {
+            repoCloner.clone(outputPath, config);
 
-            try {
-                FileUtil.deleteDirectory(FileUtil.REPOS_ADDRESS);
-            } catch (IOException ioe) {
-                logger.log(Level.WARNING, "Error deleting report directory.", ioe);
+            if (clonedRepo != null) {
+                analyzeRepo(outputPath, clonedRepo);
             }
+            clonedRepo = repoCloner.getClonedRepo(outputPath);
         }
-        FileUtil.writeJsonFile(new SummaryReportJson(configs, generationDate), getSummaryResultPath(outputPath));
-        logger.info("The report is generated at " + outputPath);
+        if (clonedRepo != null) {
+            analyzeRepo(outputPath, clonedRepo);
+        }
```

Comment:
Why is this block repeated?

```
if (clonedRepo != null) {
    analyzeRepo(outputPath, clonedRepo);
}
```
Is it possible to restructure the control flow?

Reply:
In the for loop, the current repo is cloned and the previous repo is analyzed. We loop for n times where n is length of configs, thats why the last repo is analyzed outside the for loop.

One way to avoid the repeated block is to loop n+1 times:
```
for (int i = 0; i <= configs.size(); i++) {
            if (i != configs.size())
                repoCloner.clone(outputPath, configs[i]);

            if (clonedRepo != null) {
                analyzeRepo(outputPath, clonedRepo);
            }
            if (i != configs.size())
                clonedRepo = repoCloner.getClonedRepo(outputPath);
        }
```